### ISSUE TITLE
parsing error: try-with-resources on instance of outer class

### ISSUE THREAD

**Issue by N1cOs** [checkstyle/checkstyle#8818]
$ javac Test.java
```
No output
```

$ cat config.xml
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name = "Checker">
	<module name="TreeWalker">
		<module name="IllegalTokenText">
		</module>
	</module>
</module>
```
$ cat Test.java
```
public class Test implements AutoCloseable {

    public static void main(String[] args) {
        var t = new Test();
        t.test();
    }

    private void test() {
        var a = new Abstract() {
            @Override
            void test() {
                try (Test.this) {
                    System.out.println("Unable to process");
                }
            }
        };
        a.test();
    }

    @Override
    public void close() {

    }
}

abstract class Abstract {
    abstract void test();
}
```

$ java $RUN_LOCALE -jar checkstyle-8.36-all.jar -c config.xml Test.java
```
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing Test.java
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:311)
	at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:221)
	at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:408)
	at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:331)
	at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:190)
	at com.puppycrawl.tools.checkstyle.Main.main(Main.java:125)
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: IllegalStateException occurred while parsing file /home/nick/Downloads/Test.java.
	at com.puppycrawl.tools.checkstyle.JavaParser.parse(JavaParser.java:120)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:149)
	at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:87)
	at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:337)
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:298)
	... 5 more
Caused by: java.lang.IllegalStateException: /home/nick/Downloads/Test.java:9:13: unexpected token: a
	at com.puppycrawl.tools.checkstyle.JavaParser$1.reportError(JavaParser.java:108)
	at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.typeDefinition(GeneratedJavaRecognizer.java:424)
	at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.compilationUnit(GeneratedJavaRecognizer.java:212)
	at com.puppycrawl.tools.checkstyle.JavaParser.parse(JavaParser.java:114)
	... 9 more
Caused by: /home/nick/Downloads/Test.java:9:13: unexpected token: a
	at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.traditionalStatement(GeneratedJavaRecognizer.java:6023)
	at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.statement(GeneratedJavaRecognizer.java:4814)
	at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.compoundStatement(GeneratedJavaRecognizer.java:4543)
	at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.field(GeneratedJavaRecognizer.java:3158)
	at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.classBlock(GeneratedJavaRecognizer.java:3415)
	at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.classDefinition(GeneratedJavaRecognizer.java:646)
	at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.typeDefinitionInternal(GeneratedJavaRecognizer.java:561)
	at com.puppycrawl.tools.checkstyle.grammar.GeneratedJavaRecognizer.typeDefinition(GeneratedJavaRecognizer.java:402)
	... 11 more
Checkstyle ends with 1 errors.
```
---------------

Expected behaviour: completion without error.

**Comment by pbludov** [comment]
@N1cOs according to the [Java specs](https://docs.oracle.com/javase/specs/jls/se15/html/jls-14.html#jls-14.20.3)
```
VariableAccess:
   ExpressionName
   FieldAccess 

ExpressionName:
  Identifier
  AmbiguousName . Identifier 

Identifier:
   IdentifierChars but not a Keyword or BooleanLiteral or NullLiteral 
```

keyword `this` cannot be used as VariableAccess. You found a bug in the Java compiler. Or in the Java Language Specification.
Anyway this feature is undocumented and can be changed in a future version of  Java compiler (although this is very unlikely).

For the [reference](https://github.com/checkstyle/checkstyle/issues/6332#issuecomment-612523168)

**Comment by Rahulkhinchi03** [comment]
Hello @pbludov @timurt @romani.
I am new to antlr. Can you give me some guidance to solve these issue?

**Comment by romani** [comment]
Please look at other issues marked with antlr label and hoi they were fixed. There is grammar file in our repo https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g
Documentation is available online on antrl2 tool , it is not antrl4.

**Comment by nmancus1** [comment]
This issue can be closed after 9.0 release:

Latest master:

```bash

[10:27:11] ➜  src /usr/lib/jvm/java-16-openjdk/bin/javac Test.java
[10:27:14] ➜  src cat Test.java
public class Test implements AutoCloseable {

    public static void main(String[] args) {
        var t = new Test();
        t.test();
    }

    private void test() {
        var a = new Abstract() {
            @Override
            void test() {
                try (Test.this) {
                    System.out.println("Unable to process");
                }
            }
        };
        a.test();
    }

    @Override
    public void close() {

    }
}

abstract class Abstract {
    abstract void test();
}

➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name = "Checker">
        <module name="TreeWalker">
                <module name="IllegalTokenText">
                </module>
        </module>
</module>      

[10:28:41] ➜  src java -jar ~/IdeaProjects/checkstyle/target/checkstyle-9.0-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
Audit done.
                                                                                                                   
```

**Comment by strkkk** [comment]
@nmancus1 thanks for update
